File: 2015/rust/src/day11.rs

extern crate fancy_regex;
fn main() {
    let input: String = String::from("hxbxwxba");
    let out1 = generate_new(&input, false);
    println!("{out1}");
    let out2 = generate_new(&out1, true);
    println!("{out2}");


}

fn increase(pw:Vec<u8>) -> Vec<u8> {
    let mut pw_inv: Vec<u8> = pw.into_iter().rev().collect();
    
    for (index, elem) in pw_inv.clone().into_iter().enumerate() {
        if elem + 1 > 122 {
            pw_inv[index] = 97;
        } else {
            pw_inv[index] = elem + 1;
            break; 
        }
    }
    pw_inv.into_iter().rev().collect()
}

fn generate_new(pw:&String, inc:bool) -> String {
    let mut pw_ascii = to_ascii_list(pw);
    if inc {
        pw_ascii = increase(pw_ascii);
    }
        
    while !(increasing_straight(&pw_ascii) & no_forbidden_letters(&pw_ascii) & two_pairs(&pw_ascii)) {
        pw_ascii = increase(pw_ascii);
    }

    to_str(pw_ascii)
}

fn to_ascii_list(password: &String) -> Vec<u8> {
    let mut pw = vec![0; password.len() as usize];
    for (index, char) in password.chars().enumerate() {
        pw[index] = char as u8
    }
   pw
}

fn to_str(password: Vec<u8>) -> String {
    let mut pw = String::new();
    for elem in password {
        pw.push(elem as char);
    }
   pw
}



fn increasing_straight(password:&Vec<u8>) -> bool {
    password.windows(3).any(|triple| {
    triple[0] + 1 == triple[1] && triple[1] + 1 == triple[2]
    })
    
}
fn no_forbidden_letters(password:&Vec<u8>) -> bool {

    if password.contains(&105) || password.contains(&108)|| password.contains(&111) {
        return false;
    }
    return true;
}

fn two_pairs(password:&Vec<u8>) -> bool {
    let mut pairs = 0;
    let mut i = 0;
    while i < password.len() - 1 {
        if password[i] == password[i + 1] {
            pairs += 1;
            i += 2; // skip the next char to avoid overlapping pairs
        } else {
            i += 1;
        }
    }
    pairs >= 2
}

File: 2015/python/day11.py
import re

def increase_straight_greater_three(pw:list[int]) -> bool:
    for index in range(len(pw)-2):
        if pw[index+2] - pw[index+1] == pw[index+1] - pw[index] == 1:
            return True
    return False


def no_mistaken_characters(pw:list[int]) -> bool:
    if 105 in pw or 111 in pw or 108 in pw:
        return False
    return True

def two_pairs(pw:list[int]) -> bool:
    txt = "-".join(map(str, pw))
    matches = re.findall(r"-([0-9]{2,3})-(\1)", txt)
    return len(matches) >= 2

def str_to_ascii_list(txt:str) -> list[int]:
    return [ord(c) for c in txt]

def ascii_list_to_str(inp: list[int]) -> str:
    return "".join([chr(i)for i in inp])

def increase(pw:list[int]) -> list[int]:
    pw_inv = pw[::-1]
    for index, elem in enumerate(pw_inv):
        if elem + 1 > 122:
            pw_inv[index] = 97
        else:
            pw_inv[index] += 1
            return pw_inv[::-1]

def generate_pw(inp:str, inc:bool=False) -> str:
    pw = str_to_ascii_list(inp)
    if inc:
        pw = increase(pw)
    while not (no_mistaken_characters(pw) and two_pairs(pw) and increase_straight_greater_three(pw)):
        pw = increase(pw)
    return ascii_list_to_str(pw)

old_pw = "hxbxwxba"
new_gen1 = generate_pw(old_pw)
print("Part 1", new_gen1)
new_gen2 = generate_pw(new_gen1, True)
print("Part 2", new_gen2)